ste/mgr-JobMgr.go

```diff
@@ -467,30 +472,48 @@ func (jm *jobMgr) ResetAllTransfersScheduled() {
 }
 
 // ReportJobPartDone is called to report that a job part completed or failed
-func (jm *jobMgr) ReportJobPartDone() uint32 {
+func (jm *jobMgr) ReportJobPartDone(progressInfo jobPartProgressInfo) {
+	jm.jobPartProgress <- progressInfo
+}
+
+func (jm *jobMgr) reportJobPartDoneHandler() {
+	var haveFinalPart bool
+	var jobProgressInfo jobPartProgressInfo
 	shouldLog := jm.ShouldLog(pipeline.LogInfo)
 
-	jobPart0Mgr, ok := jm.jobPartMgrs.Get(0)
-	if !ok {
-		jm.Panic(fmt.Errorf("Failed to find Job %v, Part #0", jm.jobID))
-	}
-	part0Plan := jobPart0Mgr.Plan()
-	jobStatus := part0Plan.JobStatus() // status of part 0 is status of job as a whole
-
-	partsDone := atomic.AddUint32(&jm.partsDone, 1)
-	// If the last part is still awaited or other parts all still not complete,
-	// JobPart 0 status is not changed (unless we are cancelling)
-	allKnownPartsDone := partsDone == jm.jobPartMgrs.Count()
-	haveFinalPart := atomic.LoadInt32(&jm.atomicFinalPartOrderedIndicator) == 1
-	isCancelling := jobStatus == common.EJobStatus.Cancelling()
-	shouldComplete := allKnownPartsDone && (haveFinalPart || isCancelling)
-	if !shouldComplete {
+	for {
+		partProgressInfo := <-jm.jobPartProgress
+		jobPart0Mgr, ok := jm.jobPartMgrs.Get(0)
+		if !ok {
+			jm.Panic(fmt.Errorf("Failed to find Job %v, Part #0", jm.jobID))
+		}
+		part0Plan := jobPart0Mgr.Plan()
+		jobStatus := part0Plan.JobStatus() // status of part 0 is status of job as a whole
+		partsDone := atomic.AddUint32(&jm.partsDone, 1)
+		atomic.AddInt32(&jobProgressInfo.atomicTransfersCompleted, partProgressInfo.atomicTransfersCompleted)
+		atomic.AddInt32(&jobProgressInfo.atomicTransfersSkipped, partProgressInfo.atomicTransfersSkipped)
+		atomic.AddInt32(&jobProgressInfo.atomicTransfersFailed, partProgressInfo.atomicTransfersFailed)
+
+		// If the last part is still awaited or other parts all still not complete,
+		// JobPart 0 status is not changed (unless we are cancelling)
+		haveFinalPart = atomic.LoadInt32(&jm.atomicFinalPartOrderedIndicator) == 1
+		allKnownPartsDone := partsDone == jm.jobPartMgrs.Count()
+		isCancelling := jobStatus == common.EJobStatus.Cancelling()
+		shouldComplete := allKnownPartsDone && (haveFinalPart || isCancelling)
+		if shouldComplete {
+			break
+		} //Else log and wait for next part to complete
+
 		if shouldLog {
 			jm.Log(pipeline.LogInfo, fmt.Sprintf("is part of Job which %d total number of parts done ", partsDone))
 		}
-		return partsDone
+
+		jm.chunkStatusLogger.FlushLog()
```

Comment:
I'd suggest removing this call to `FlushLog`, and just leave the one call at the end of reportJobPartDoneHandler.  Why? Because this Flush() function has a sticky mode: after its been called once, it always flushes _after every single write_ from then on. Which could have a perf cost. So Flush() should only be called once at the end of the job.

(Admittedly, the sticky flush mode may not have been my best design decision!  But that's the way it works for now).

Reply:
Done